### ISSUE TITLE
bug(python): btree index search is very slow

### ISSUE THREAD

**Issue by oswardlx** [lancedb/lancedb#1734]
### LanceDB version

0.13.0

### What happened?

I have 140 million data, one of which is a 1024 dimensional vector and the other is a float type. I have established an IVF-PQ index for the vector column and a BTREE index for the float column. Vector column queries can quickly recall limit 20 in about 300ms, but when I perform vector queries and also pre filter queries on the float column, such as price>10 and price<30, limit 20, it takes 55 seconds. Is this a normal phenomenon? Can this mixed recall only be so slow? Is there a way to speed it up

**Comment by oswardlx** [comment]
this my search code i used sync connect and table
gogo = tbl.search(np.random.random(1024), 'embedding').where("price < 30 and price >10",prefilter=True).limit(20).to_pandas().drop(columns='embedding').to_json(orient='records', force_ascii=False)

**Comment by wjones127** [comment]
How selective is your filter? You could compute that with:

```python
selected_rows = tbl.count_rows("price < 30 and price >10")
total_rows = tbl.count_rows()
print(selected_rows / total_rows
```

If your filter is very selective, it might not be very efficient to use it with a vector index. How prefiltering works is that the ANN index provides candidate vectors, and then we check whether they match the filter and throw out results that don't. We keep doing this until we have enough results. If the filter is very selective, then you'll end up spending most of the time throwing out candidates found in your ANN index. There are other ways to query that are more efficient in that case. Though not sure if that's what you are hitting.

**Comment by oswardlx** [comment]
> How selective is your filter? You could compute that with:
> 
> ```python
> selected_rows = tbl.count_rows("price < 30 and price >10")
> total_rows = tbl.count_rows()
> print(selected_rows / total_rows
> ```
> 
> If your filter is very selective, it might not be very efficient to use it with a vector index. How prefiltering works is that the ANN index provides candidate vectors, and then we check whether they match the filter and throw out results that don't. We keep doing this until we have enough results. If the filter is very selective, then you'll end up spending most of the time throwing out candidates found in your ANN index. There are other ways to query that are more efficient in that case. Though not sure if that's what you are hitting.

It doesn't seem to be a matter of quantity,
When I use conditions
gogo = tbl.search(zz, 'embedding').where("price >= 29.9 and price<=30",prefilter=True).limit(3).to_pandas().to_json(orient='records', force_ascii=False)
search data executed in: 58.992310 seconds
select rows:991359
total_rows:148144132
When I use:
gogo = tbl.search(zz, 'embedding').where("price=30",prefilter=True).limit(3).to_pandas().to_json(orient='records', force_ascii=False)
search data executed in: 0.270848 seconds!!!!
select rows:836336
total_rows:148144132
The time difference is very large, it seems that the range query time consumption will be greater. Can this be solved

**Comment by oswardlx** [comment]
price >= 29.999 and price = 30
search data executed in: 58.508147 seconds
836336
148144132

The quantity above is the same as price=30, but there is still a significant difference in time

**Comment by wjones127** [comment]
That's 0.5% selectivity. Which means it has to find 200x more results in the index. So it might not be surprising that it takes this long.

**Comment by westonpace** [comment]
Yes, @wjones127 is right.  We can, and should, be smarter about this too.

An indexed search currently translates into random access.  Random access of 200k rows is slow.  This is especially true on cloud storage (are you using cloud storage?).  Although it does slightly depend on the column.

We should detect when an indexed search has returned too many results and fall back to a "full scan and throw away rows we don't need" instead of "random access" (the math is straightforward and based on "is this cloud or local", "is this a scalar type or a vector/image/binary type" and "how many row ids were returned from the index search").

This is a related issue: https://github.com/lancedb/lance/issues/1896

**Comment by oswardlx** [comment]
I'm not sure if I made myself clear. When I use conditions 1: price>=29.999 and price<=30 respectively. Condition 2: Price=30. Their count_rows have the same quantity of 836336
But condition 1 recall takes 58 seconds, while condition 2 recall only takes 0.27 seconds. There is a significant difference, so this should have nothing to do with 0.5% selectivity, right. Also, I am using a local disk, and both of these conditions are recalled using the same vector, and both are prefilter=Ture

@westonpace @wjones127

**Comment by westonpace** [comment]
Ah,  you're right, I didn't read carefully enough.

Can you run the slow query with tracing turned on and upload the trace file?

Just add this to the top of your python script:

```
from lance.tracing import trace_to_chrome
trace_to_chrome(file="/some/path/trace.json", level="debug")
```

Once your script is done running then `/some/path/trace.json` should be created.  If you upload it then we can take a closer look.  If you want to analyze the trace yourself you can also do so at ui.perfetto.dev

**Comment by oswardlx** [comment]
[trace.tar.gz](https://github.com/user-attachments/files/17362744/trace.tar.gz)

ui.perfetto.dev has some problem

**Comment by westonpace** [comment]
Hmm, there is a lot of I/O happening which shouldn't be needed for a prefiltered vector search.  Can you run compaction and update your indices?  With `lancedb` this should be a call to https://lancedb.github.io/lancedb/python/python/#lancedb.table.AsyncTable.optimize

If you're using the synchronous python API you can do this today by dropping into pylance...

```
# For sync API, compact & update
tbl.compact_files()
tbl.to_lance().optimize.optimize_indices()
```
